mlir/lib/Transforms/NormalizeMemRefs.cpp

```diff
@@ -0,0 +1,221 @@
+//===- NormalizeMemRefs.cpp ---------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements an interprocedural pass to normalize memref layouts.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "mlir/Analysis/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/IR/Function.h"
+#include "mlir/IR/FunctionSupport.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/Utils.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/Format.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "normalize-memrefs"
+
+using namespace mlir;
+
+namespace {
+
+/// All interprocedural memrefs with non-trivial layout maps are converted to
+/// ones with trivial identity layout ones.
+
+// Input :-
+// #tile = affine_map<(i) -> (i floordiv 4, i mod 4)>
+// func @matmul(%A: memref<16xf64, #tile>, %B: index, %C: memref<16xf64>) ->
+// (memref<16xf64, #tile>) {
+//   affine.for %arg3 = 0 to 16 {
+//         %a = affine.load %A[%arg3] : memref<16xf64, #tile>
+//         %p = mulf %a, %a : f64
+//         affine.store %p, %A[%arg3] : memref<16xf64, #tile>
+//   }
+//   %c = alloc() : memref<16xf64, #tile>
+//   %d = affine.load %c[0] : memref<16xf64, #tile>
+//   return %A: memref<16xf64, #tile>
+// }
+
+// Output :-
+// module {
+//   func @matmul(%arg0: memref<4x4xf64>, %arg1: index, %arg2: memref<16xf64>)
+//   -> memref<4x4xf64> {
+//     affine.for %arg3 = 0 to 16 {
+//       %2 = affine.load %arg0[%arg3 floordiv 4, %arg3 mod 4] : memref<4x4xf64>
+//       %3 = mulf %2, %2 : f64
+//       affine.store %3, %arg0[%arg3 floordiv 4, %arg3 mod 4] : memref<4x4xf64>
+//     }
+//     %0 = alloc() : memref<16xf64, #map0>
+//     %1 = affine.load %0[0] : memref<16xf64, #map0>
+//     return %arg0 : memref<4x4xf64>
+//   }
+// }
+
+struct NormalizeMemRefs : public NormalizeMemRefsBase<NormalizeMemRefs> {
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+std::unique_ptr<OperationPass<ModuleOp>> mlir::createNormalizeMemRefsPass() {
+  return std::make_unique<NormalizeMemRefs>();
+}
+
+void NormalizeMemRefs::runOnOperation() {
+  ModuleOp moduleOp = getOperation();
+
+  moduleOp.walk([&](FuncOp funcOp) {
```

Comment:
I didn't notice this earlier: you could just outline this lambda. Please create a `runOnFunction(FuncOp funcOp)` and outline this out. A lambda function this long should really be a separate function for better readability. You also get rid of the extra indent/nesting that way.

Reply:
Have outlined the lambda function into runOnFunction(FuncOp funcOp).
How to refactor the code so that AllocOp and a function argument can use the same code needs a little thought. I'd look into that for sure because the readability has indeed improved after the said changes and I'm looking forward to bettering it.

Reply:
> Have outlined the lambda function into runOnFunction(FuncOp funcOp).
> How to refactor the code so that AllocOp and a function argument can use the same code needs a little thought. I'd look into that for sure because the readability has indeed improved after the said changes and I'm looking forward to bettering it.

Sounds great. You can start with moving that `normalizeMemRef` in this file (right next to runOnFunction) and we can see how to refactor it to share code. It should be straightforward if we take a crack.

Reply:
Sure, sir. Will look into this.
HigherOrderPolyhedralOpt.cpp uses `normalizeMemRef`, will have to change the the path and the namespace.

Reply:
> Sure, sir. Will look into this.
> `HigherOrderPolyhedralOpt.cpp` uses `normalizeMemRef`, will have to change the the path and the namespace.

I see - good catch. Note that HOPT doesn't exist upstream. We may still want to have that callable utility - we need to see how to refactor it so that it can be used from the pass as well as separately for alloc ops.